### PR TITLE
KBinsDiscretizer Aggregate Encoder (Helper + Mode) #19081

### LINKED ISSUE #6: Implement Mode encoder for KBinsDiscretizer

### LINKED ISSUE #9: Implement encoder helper for KBinsDiscretizer

### PR DESCRIPTION
**Pull request by RylandWang**
#### Reference Issues/PRs
References #19081. 

#### What does this implement/fix? Explain your changes.
This extends the functionality of KBinsDiscretizer to encode matrices with aggregate functions such as min, max, mean, and mode.
I created a higher order helper function that utilizes ordinal encoding to bin data into discrete intervals, and simply requires a lambda function like min or max to be passed in.

#### Example
image

#### output
image

### CODE REVIEW

File: sklearn/preprocessing/_discretization.py [hunk 1]
```diff
@@ -351,3 +361,30 @@ def inverse_transform(self, Xt):
             Xinv[:, jj] = bin_centers[np.int_(Xinv[:, jj])]
 
         return Xinv
+
+
+    def _aggregate_encoder_helper(self, X, aggregate_encoder=lambda x: np.min(x)):
+        A = np.array(X)
+
+        ordinal_labels = KBinsDiscretizer(self.n_bins, encode='ordinal').fit_transform(A)
```
Comment (linjun9) [inline sklearn/preprocessing/_discretization.py]: I think we already have the labels ready,  no need to fit_transform again.

File: sklearn/preprocessing/_discretization.py [hunk 2]
```diff
@@ -303,6 +305,14 @@ def transform(self, X):
             Xt[:, jj] = np.digitize(Xt[:, jj] + eps, bin_edges[jj][1:])
         np.clip(Xt, 0, self.n_bins_ - 1, out=Xt)
 
+        if self.encode == 'max':
+            encoder = lambda x: np.max(x)
+            return self._aggregate_encoder_helper(X, encoder)
```
Comment (linjun9) [inline sklearn/preprocessing/_discretization.py]: we can pass the labels to this function, you can get the labels by `labels = Xt.astype(int)`

File: sklearn/preprocessing/_discretization.py [hunk 3]
```diff
@@ -351,3 +361,30 @@ def inverse_transform(self, Xt):
             Xinv[:, jj] = bin_centers[np.int_(Xinv[:, jj])]
 
         return Xinv
+
+
+    def _aggregate_encoder_helper(self, X, aggregate_encoder=lambda x: np.min(x)):
+        A = np.array(X)
```
Comment (linjun9) [inline sklearn/preprocessing/_discretization.py:368]: not sure if this is the best way to create an `array` with the same dimension? I think `np.zeros_like` will be a better choice? it filled the array with zeros not creating a copy of the original array.

File: sklearn/preprocessing/tests/test_discretization.py [hunk 4]
```diff
@@ -207,6 +207,39 @@ def test_nonuniform_strategies(
...
 
 
+def test_transform_mode():
+    X = [[-2, 1, -1,   -1],
+        [-1, 2, -3,   0.5],
+        [ 0, 4, -2,  0.5],
+        [ 1, 4, -1,    2],
+        [ 1,  5,  0,   2]]
+    
+    expected_2bins = [[-2, 1, -1,  -1],
```
Comment (linjun9) [inline sklearn/preprocessing/tests/test_discretization.py:217]: It's better to separate this test case into three.

**Review by linjun9 — COMMENTED** [review]
@RylandWang Can you also attach the screenshot of the test result, because the pipeline doesn't run for our repo.

File: sklearn/preprocessing/tests/test_discretization.py (hunk 4 above)
Comment (RylandWang) [reply sklearn/preprocessing/tests/test_discretization.py:217]: I'm following the convention of the existing test cases. It seems like they had multiple assertions with different bins sizes in a single test case.

File: sklearn/preprocessing/_discretization.py (hunk 3 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py:368]: Do you think there is a complexity difference between the two?

File: sklearn/preprocessing/_discretization.py (hunk 1 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py]: Oh, you're correct.

File: sklearn/preprocessing/_discretization.py (hunk 3 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py:368]: For some reason, I couldn't get it to work with np.zeros_like

Comment (linjun9) [reply sklearn/preprocessing/_discretization.py:368]: yea, I thought there would be a complexity difference at first. as you need to go through the original array, and assign the values to elements in the new array. But now I look at it, that may be wrong. you also need to assign zeros to the new array anyways.

File: sklearn/preprocessing/tests/test_discretization.py (hunk 4 above)
Comment (linjun9) [reply sklearn/preprocessing/tests/test_discretization.py:217]: yea, you are right. I guess that's fine then.

File: sklearn/preprocessing/_discretization.py (hunk 3 above)
Comment (linjun9) [reply sklearn/preprocessing/_discretization.py:368]: I'm ok with either one.

Comment (linjun9) [reply sklearn/preprocessing/_discretization.py:368]: > For some reason, I couldn't get it to work with np.zeros_like

maybe more info about this?

Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py:368]: Never mind, np.zeros_like works as well. But I'm still using my original method since I don't see any difference between the two.

File: sklearn/preprocessing/_discretization.py (hunk 1 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py]: Updated.

File: sklearn/preprocessing/_discretization.py [hunk 5]
```diff
@@ -1,353 +1,389 @@
...
+
+        for jj in range(n_features):
+            bin_edges = self.bin_edges_[jj]
+            bin_centers = (bin_edges[1:] + bin_edges[:-1]) * 0.5
+            Xinv[:, jj] = bin_centers[np.int_(Xinv[:, jj])]
+
+        return Xinv
+
+
+    def _aggregate_encoder_helper(self, X, ordinal_labels, aggregate_encoder=lambda x: np.min(x)):
```
Comment (linjun9) [inline sklearn/preprocessing/_discretization.py]: I don't think it's a good idea to set the default encoder to `min`? Is there a reason for that?

**Review by linjun9 — COMMENTED** [review]
There are so many lines changed. I couldn't tell the difference. Did you change the format or something?

File: sklearn/preprocessing/_discretization.py [hunk 6]
```diff
@@ -1,353 +1,389 @@
...
+        # traverse columns
+        for j in range(len(A[0])):
+            col = A[:, j]
+
+            # go through each bin to get the entries that match the corresponding label
+            # then perform the encoding operation (min, max, mean, etc.) on just those entries
+            for bin_num in range(self.n_bins):
+                mapping = [] # map for membership in current bin
+                # traverse labels in this column
+                for i in range(len(ordinal_labels[:, j])):
```
Comment (VishwaP98) [inline sklearn/preprocessing/_discretization.py]: @RylandWang we can just replace the len(ordinal_labels[:, j]) with len(A) since we are traversing through all the row values

File: sklearn/preprocessing/_discretization.py [hunk 7]
```diff
@@ -1,353 +1,389 @@
...
+        A = np.array(X)
+
+        # traverse columns
+        for j in range(len(A[0])):
+            col = A[:, j]
+
+            # go through each bin to get the entries that match the corresponding label
+            # then perform the encoding operation (min, max, mean, etc.) on just those entries
+            for bin_num in range(self.n_bins):
+                mapping = [] # map for membership in current bin
```
Comment (VishwaP98) [inline sklearn/preprocessing/_discretization.py]: @RylandWang  Initialize an array of len(A) with False and only set the value to True if the rowItem is a part of the bin

File: sklearn/preprocessing/_discretization.py [hunk 8]
```diff
@@ -1,353 +1,389 @@
...
+                # traverse labels in this column
+                for i in range(len(ordinal_labels[:, j])):
+                    # member
+                    if ordinal_labels[i, j] == bin_num:
+                        mapping.append(True)
+                    # not memeber
+                    else:
```
Comment (VishwaP98) [inline sklearn/preprocessing/_discretization.py]: This will not be needed if we initialize before

File: sklearn/preprocessing/_discretization.py [hunk 9]
```diff
@@ -1,353 +1,389 @@
...
+                        mapping.append(False)
+                if len(col[mapping]) > 0:
```
Comment (VishwaP98) [inline sklearn/preprocessing/_discretization.py]: @RylandWang  adding a comment here about what col[mapping] does can make it easier for others to understand (just a small comment about how the mapping helps in picking the relevant items from col array based on the bool value)

File: sklearn/preprocessing/_discretization.py (hunk 6 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py]: If they both accomplish the same thing, what difference does it make?

File: sklearn/preprocessing/_discretization.py (hunk 7 above)
Comment (RylandWang) [reply sklearn/preprocessing/_discretization.py]: This is slightly less efficient than my method.

Comment (VishwaP98) [reply sklearn/preprocessing/_discretization.py]: it makes the code simpler and does not have a huge impact on the efficiency

File: sklearn/preprocessing/_discretization.py (hunk 6 above)
Comment (VishwaP98) [reply sklearn/preprocessing/_discretization.py]: with len(ordinal_labels[:, j]) you are first of all accessing the entire row for a particular column and then calculating the number of rows whereas accessing len(A) will be a constant operation. len(A) is much more efficient than len(ordinal_labels[:, j]) especially for the large matrices.

**Comment by RylandWang** [comment]
> @RylandWang Can you also attach the screenshot of the test result, because the pipeline doesn't run for our repo.

image

**Comment by RylandWang** [comment]
> There are so many lines changed. I couldn't tell the difference. Did you change the format or something?

Sorry, it was a line break issue when I switched from Linux to Windows.

**Review by VishwaP98 — APPROVED** [review]
looks good to me!

**Review by linjun9 — APPROVED** [review]
👍
